test-data/unit/check-type-aliases.test

```diff
@@ -796,3 +796,154 @@ S = TypeVar("S")
+def weird_child_3() -> c.SpecialImplicit: pass  # E: Name "c.SpecialImplicit" is not defined
+def weird_child_4() -> c.SpecialExplicit: pass  # E: Name "c.SpecialExplicit" is not defined
+reveal_type(weird_child_1)  # N: Revealed type is "def () -> Any"
+reveal_type(weird_child_2)  # N: Revealed type is "def () -> Any"
+reveal_type(weird_child_3)  # N: Revealed type is "def () -> Any"
+reveal_type(weird_child_4)  # N: Revealed type is "def () -> Any"
+[builtins fixtures/tuple.pyi]
+
+[case testMalformedTypeAliasRuntimeReassignments]
+from typing import Union
+from typing_extensions import TypeAlias
+
+class Foo: pass
+
+NormalImplicit = Foo
+NormalExplicit: TypeAlias = Foo
+SpecialImplicit = Union[int, str]
+SpecialExplicit: TypeAlias = Union[int, str]
+
+class Parent:
+    NormalImplicit = Foo
+    NormalExplicit: TypeAlias = Foo
+    SpecialImplicit = Union[int, str]
+    SpecialExplicit: TypeAlias = Union[int, str]
+
+class Child(Parent): pass
+
+p = Parent()
+c = Child()
+
+NormalImplicit = 4   # E: Cannot assign multiple types to name "NormalImplicit" without an explicit "Type[...]" annotation \
+                     # E: Incompatible types in assignment (expression has type "int", variable has type "Type[Foo]")
+NormalExplicit = 4   # E: Cannot assign multiple types to name "NormalExplicit" without an explicit "Type[...]" annotation \
+                     # E: Incompatible types in assignment (expression has type "int", variable has type "Type[Foo]")
+SpecialImplicit = 4  # E: Cannot assign multiple types to name "SpecialImplicit" without an explicit "Type[...]" annotation
+SpecialExplicit = 4  # E: Cannot assign multiple types to name "SpecialExplicit" without an explicit "Type[...]" annotation
+
+Parent.NormalImplicit = 4  # E: Incompatible types in assignment (expression has type "int", variable has type "Type[Foo]")
+Parent.NormalExplicit = 4  # E: Incompatible types in assignment (expression has type "int", variable has type "Type[Foo]")
+Parent.SpecialImplicit = 4
+Parent.SpecialExplicit = 4
+
+Child.NormalImplicit = 4  # E: Incompatible types in assignment (expression has type "int", variable has type "Type[Foo]")
+Child.NormalExplicit = 4  # E: Incompatible types in assignment (expression has type "int", variable has type "Type[Foo]")
+Child.SpecialImplicit = 4
+Child.SpecialExplicit = 4
+
+p.NormalImplicit = 4  # E: Incompatible types in assignment (expression has type "int", variable has type "Type[Foo]")
+p.NormalExplicit = 4  # E: Incompatible types in assignment (expression has type "int", variable has type "Type[Foo]")
+p.SpecialImplicit = 4
+p.SpecialExplicit = 4
+
+c.NormalImplicit = 4  # E: Incompatible types in assignment (expression has type "int", variable has type "Type[Foo]")
+c.NormalExplicit = 4  # E: Incompatible types in assignment (expression has type "int", variable has type "Type[Foo]")
+c.SpecialImplicit = 4
+c.SpecialExplicit = 4
```

Comment:
Letting users modify type aliases like this isn't a great user experience IMO, but it looks like this behavior predates my diff.

One possible solution might be to make mypy give Union/Literal/Callable/etc expressions a runtime type of `typing._SpecialForm` instead of `builtins.object`? I can try submitting a PR to test that out after this one lands.

Reply:
Using `typing._SpecialForm` instead of `object` sounds like a good idea.